### PR TITLE
[CP-3166] Enhance Sorting Logic with Grouping, Null Handling, and Case Insensitivity

### PR DESCRIPTION
**Pull request by dkarski**
JIRA Reference: [CP-3166]

### :memo: Description ️

-

### :muscle: Checklist before requesting a review - nice to have

- getState function in async thunk actions is correctly typed
- redux selectors are used in components / prop drilling is reduce

### :exclamation: Checklist before merging a pull request

- change went through the QA process
- translations are updated in dedicated application
- [CHANGELOG.md](./CHANGELOG.md) is updated

[CP-3166]: https://appnroll.atlassian.net/browse/CP-3166?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: libs/generic-view/utils/src/lib/data-provider-helpers/data-provider-sort.ts [hunk 1]
```diff
@@ -4,55 +4,79 @@
...
-          direction === "asc"
-            ? fieldA.localeCompare(fieldB)
-            : fieldB.localeCompare(fieldA)
-        if (score !== 0) {
-          break
+
+        if (typeof fieldA !== "string" || typeof fieldB !== "string") {
+          index++
+          continue
+        }
```
Comment (mkurczewski) [inline libs/generic-view/utils/src/lib/data-provider-helpers/data-provider-sort.ts:54]: Tests aren't covering this lines. In fact, you can remove them and tests will pass anyway. I'd consider removing it or adding more test cases that will handle such scenario.

File: libs/device/models/src/lib/feature/data-provider-config.ts [hunk 2]
```diff
@@ -64,14 +64,37 @@ export type DataProviderField =
   | z.infer<typeof enhancedFieldSchema>
   | z.infer<typeof superEnhancedFieldSchema>
 
+const sortDirectionSchema = z.union([z.literal("asc"), z.literal("desc")])
+
+export type sortDirection = z.infer<typeof sortDirectionSchema>
```
[the remark this answers is not in the document]
Comment (dkarski) [reply libs/device/models/src/lib/feature/data-provider-config.ts]: https://github.com/mudita/mudita-center/pull/2116/commits/d6be94b3fea35840c7726886d8e3fa55e420286a

File: libs/generic-view/utils/src/lib/data-provider-helpers/data-provider-sort.ts (hunk 1 above)
Comment (dkarski) [reply libs/generic-view/utils/src/lib/data-provider-helpers/data-provider-sort.ts]: https://github.com/mudita/mudita-center/pull/2116/commits/d806f4322bb0d1e769f9b001ee281253b943ad07

**Review by mkurczewski — APPROVED** [review]

**Comment by dkarski** [comment]
I'm marking the PR as a draft due to the missing refactor of the providerField and providerFieldGroup variables to field and fieldGroup, in order to eliminate the keyword "provider" from the helpers for sort/filter.

**Review by mkurczewski — APPROVED** [review]

**Comment by dkarski** [comment]
duplicate with https://github.com/mudita/mudita-center/pull/2146

**Pull request merged**